### PR TITLE
feat: (REPLAT-6797) add viewport tracking for unruly ads

### PR DESCRIPTION
**Pull request by jmars**
This ads viewport tracking for unruly so that they can know when they are in view before starting or stopping a video ad.

I am not sure of a good way to test this? Any suggestions?

Also currently waiting for an approved list of device data to pass to them without violating user data confidentiality.

### CODE REVIEW

**Comment by nathggns** [comment]
Might be worth using something like https://blog.netcetera.com/viewport-aware-components-for-react-native-apps-a28ea605e89e instead of a timer.

File: packages/ad/src/dom-context.js [hunk 1]
```diff
@@ -55,6 +63,37 @@ class DOMContext extends PureComponent {
     }
   };
 
+  startWatchingViewport() {
+    if (this.interval) {
+      return;
+    }
+    this.interval = setInterval(() => {
```
Comment (parkjh) [inline packages/ad/src/dom-context.js]: is this the only possible way?

File: packages/ad/src/dom-context.js [hunk 2]
```diff
@@ -135,7 +175,17 @@ class DOMContext extends PureComponent {
             baseUrl,
             html
           }}
-          style={{ backgroundColor: "transparent" }}
+          style={{ position: "absolute" }}
+        />
+        <ViewportAwareView
+          onViewportEnter={() => this.inViewport()}
+          onViewportLeave={() => this.outViewport()}
```
Comment (ValGeorgiev) [inline packages/ad/src/dom-context.js]: Could we not use arrow functions in render, please use arrow functions in the declaration of inViewport and outViewport or bind them in the constructor

File: packages/ad/src/dom-context.js [hunk 3]
```diff
@@ -135,7 +175,17 @@ class DOMContext extends PureComponent {
...
+          style={{
+            width: 10,
+            height: height / 10,
+            top: height / 2 - height / 20,
```
Comment (ValGeorgiev) [inline packages/ad/src/dom-context.js]: could we use constants for these magic numbers

**Comment by ValGeorgiev** [comment]
@jmars Nice one 🚀 Could you rebase this one and check my comments. Lets merge this, because  its here for a while :)

**Comment by parkjh** [comment]
@jmars closing due to inactivity

**Comment by ValGeorgiev** [comment]
> @jmars closing due to inactivity

@parkjh @jmars rebased 18h ago ???

**Comment by parkjh** [comment]
@ValGeorgiev oops didn't see that... but to be fair this pr has been around for ages...

**Comment by jmars** [comment]
Yes it got put aside for a while due to other things needed for release, fixing up Vals comments now

**Comment by parkjh** [comment]
@jmars are there no additional functional tests? only the snapshots are updated...

**Pull request merged**